src/mapping/parser/mod.rs

```diff
@@ -195,151 +202,134 @@ fn query_arithmetic_from_pair(pair: Pair<Rule>) -> Result<Box<dyn query::Functio
-        }
-        Rule::uuid_v4 => Ok(Box::new(UuidV4Fn::new())),
-        Rule::sha1 => {
-            let pair = pair.into_inner().next().ok_or(TOKEN_ERR)?;
-            let query = query_arithmetic_from_pair(pair)?;
-            Ok(Box::new(Sha1Fn::new(query)))
-        }
-        Rule::md5 => {
-            let pair = pair.into_inner().next().ok_or(TOKEN_ERR)?;
-            let query = query_arithmetic_from_pair(pair)?;
-            Ok(Box::new(Md5Fn::new(query)))
-        }
-        Rule::now => Ok(Box::new(NowFn::new())),
-        Rule::strip_whitespace => {
-            let param = pair.into_inner().next().ok_or(TOKEN_ERR)?;
-            let query = query_arithmetic_from_pair(param)?;
-            Ok(Box::new(StripWhitespaceFn::new(query)))
-        }
-        Rule::truncate => {
-            let (first, mut other) = split_inner_rules_from_pair(pair)?;
-            let query = query_arithmetic_from_pair(first)?;
-            let limit = query_arithmetic_from_pair(other.next().ok_or(TOKEN_ERR)?)?;
+fn query_function_from_pairs(mut pairs: Pairs<Rule>) -> Result<Box<dyn query::Function>> {
+    let name = pairs.next().ok_or(TOKEN_ERR)?.as_span().as_str();
+    let signature = FunctionSignature::from_str(name)?;
+    let arguments = function_arguments_from_pairs(pairs, &signature)?;
 
-            let ellipsis = other.next().map(|r| match r.as_rule() {
-                Rule::boolean => Value::Boolean(r.as_str() == "true"),
-                Rule::null => Value::Null,
-                _ => unreachable!(
-                    "parser should not allow other to_bool default arg child rules here"
-                ),
-            });
+    signature.into_boxed_function(arguments)
+}
 
-            Ok(Box::new(TruncateFn::new(query, limit, ellipsis)))
-        }
-        Rule::parse_json => {
-            let param = pair.into_inner().next().ok_or(TOKEN_ERR)?;
-            let query = query_arithmetic_from_pair(param)?;
-            Ok(Box::new(ParseJsonFn::new(query)))
-        }
+fn function_arguments_from_pairs(
+    mut pairs: Pairs<Rule>,
+    signature: &FunctionSignature,
+) -> Result<ArgumentList> {
+    let mut arguments = ArgumentList::new();
+
+    // Check if any arguments are provided.
+    if let Some(pairs) = pairs.next().map(|pair| pair.into_inner()) {
+        // Keeps track of positional argument indices.
+        //
+        // Used to map a positional argument to its keyword. Keyword arguments
+        // can be used in any order, and don't count towards the index of
+        // positional arguments.
+        let mut index = 0;
 
-        _ => unreachable!("parser should not allow other query_function child rules here"),
+        pairs
+            .map(|pair| pair.into_inner().next().unwrap())
+            .map(|pair| match pair.as_rule() {
+                Rule::positional_item => {
+                    index += 1;
+                    positional_item_from_pair(pair, &mut arguments, index - 1, signature)
```

Comment:
Could you increment the `index` after this call rather than before? Then you can just pass `index` rather than `index - 1`

Reply:
This is in a `map()` call, so to do that, I'd have to capture the output first, then add 1 to `index`, and then return the captured value from `positional_item_from_pair`. That seemed more complicated than it's worth to prevent the `-1`. Or do you believe, given this, it's still worth doing?

Reply:
Oh, duh. No, sorry I missed that. Your way is much better. Ignore me here..